Modules/_threadmodule.c

```diff
@@ -1217,26 +1243,55 @@ thread_PyThread_start_new_thread(PyObject *self, PyObject *fargs)
     boot->args = Py_NewRef(args);
     boot->kwargs = Py_XNewRef(kwargs);
 
-    unsigned long ident = PyThread_start_new_thread(thread_run, (void*) boot);
+    unsigned long ident;
+    Py_uintptr_t handle = 0;
+    if (joinable == Py_True) {
+        ident = PyThread_start_joinable_thread(thread_run, (void*) boot, &handle);
+    } else {
+        ident = PyThread_start_new_thread(thread_run, (void*) boot);
+    }
     if (ident == PYTHREAD_INVALID_THREAD_ID) {
         PyErr_SetString(ThreadError, "can't start new thread");
         PyThreadState_Clear(boot->tstate);
         thread_bootstate_free(boot, 1);
         return NULL;
     }
-    return PyLong_FromUnsignedLong(ident);
+    PyObject* ident_obj = PyLong_FromUnsignedLong(ident);
+    if (ident_obj == NULL) {
+        return NULL;
+    }
+    if (joinable == Py_True) {
+        PyObject* handle_obj = PyLong_FromUnsignedLongLong(handle);
+        if (handle_obj == NULL) {
+            Py_DECREF(ident_obj);
+            return NULL;
+        }
+        if (PyDict_SetItem(state->joinable_dict, ident_obj, handle_obj)) {
+            Py_DECREF(handle_obj);
+            Py_DECREF(ident_obj);
+            return NULL;
+        }
+        Py_DECREF(handle_obj);
+    }
+    return ident_obj;
```

Comment:
IIUC, in the error cases here (i.e. post-start), the thread will keep running.  That might be confusing for users and might also interfere with the operation of the whole `threading.Thread.start()` call stack.  Is there something we could do to avoid that situation?

Also, in each of these cases, we should be careful about ensuring the thread state and bootstate are cleared/freed.

<details>
<summary>Some thoughts:</summary>

Without having given it a lot of thought, there seem to be two high-level options: avoid any possible errors once we've started the thread, and make a best effort at stopping the thread.  The latter doesn't seem feasible.

However, the former seems possible.  We can avoid errors after starting the thread by identifying how all post-start operations might fail and moving such failure paths to before we start the thread.

For example:

```C
    PyObject *ident_obj = NULL;
    PyObject *handle_obj = NULL;
    // The thread state ID is just as unique as the OS thread ID.
    PyObject *tsid = PyLong_FromUnsignedLongLong(boot->tstate->id);
    if (tsid == NULL) {
        goto error;
    }
    // Preallocate, to avoid failures after starting the thread.
    ident_obj = _PyLong_NEW_ULLONG();
    if (ident_obj == NULL) {
        goto error;
    }
    Py_uintptr_t handle = 0;
    if (joinable == Py_True) {
        // Preallocate and add to the joinables, to avoid failures after starting the thread.
        handle_obj = _PyLong_NEW_ULLONG();
        if (handle_obj == NULL) {
             goto error;
        }
        if (PyDict_SetItem(state->joinable_dict, tsid, handle_obj)) {
            goto error;
        }

        ident = PyThread_start_joinable_thread(thread_run, (void*) boot, &handle);

        if (ident == PYTHREAD_INVALID_THREAD_ID) {
            if (PyDict_DelItem(state->joinable_dict, tsid)) {
                PyErr_Clear();  // ???
            }
        } else {
            _PyLong_SetValue(ident_obj, ident);
            _PyLong_SetValue(handle_obj, handle);
        }
        Py_CLEAR(handle_obj);
    } else {
        ident = PyThread_start_new_thread(thread_run, (void*) boot);
    }
    if (ident == PYTHREAD_INVALID_THREAD_ID) {
        PyErr_SetString(ThreadError, "can't start new thread");
        goto error;
    }
    return ident_obj;

error:
    Py_XDECREF(tsid);
    Py_XDECREF(ident_obj);
    Py_XDECREF(handle_obj);
    PyThreadState_Clear(boot->tstate);
    thread_bootstate_free(boot, 1);
    return NULL;
```

That really isn't the best implementation, but hopefully it demonstrates the idea.

Note that `_PyLong_NEW_ULLONG()` and `_PyLong_SetValue()` don't currently exist (and might actually be a bad idea).  The former is equivalent to `_PyLong_FromUnsignedLongLong(ULLONG_MAX)` (or just `_PyLong_New((sizeof(unsigned long long)+PyLong_SHIFT-1)/PyLong_SHIFT)`).  For the latter, imagine it is basically the last few lines of PYLONG_FROM_UINT() in longobject.c.

Again, this all might be simpler if the handle were a struct with the actual threadid/handle (and other state) in it, and were initialized in `PyThread_start_joinable_thread()`.

```C
    _PyThreadIDObject *ident_obj = _PyThreadID_New();
    if (ident_obj == NULL) {
        goto error;
    }
    int res;
    if (joinable == Py_True) {
        res = PyThread_start_joinable_thread(thread_run, (void*) boot, &ident_obj->handle);
    } else {
        unsigned long ident = PyThread_start_new_thread(thread_run, (void*) boot);
        res = (ident == PYTHREAD_INVALID_THREAD_ID) ? -1 : 0;
        ident_obj->handle.ident = ident;
    }
    if (res < 0) {
        PyErr_SetString(ThreadError, "can't start new thread");
        goto error;
    }
    return ident_obj;

error:
    Py_XDECREF(ident_obj);
    PyThreadState_Clear(boot->tstate);
    thread_bootstate_free(boot, 1);
    return NULL;
```

</details>

Reply:
This is addressed now, as the thread handle object that's returned is allocated before starting the thread.